File: packages/domain-floorplan/src/measurement.test.ts
import { describe, expect, it } from "vitest";
import { DomainError } from "./domain-error.js";
import { FloorPlanDocument } from "./floor-plan-document.js";
import { LengthMm } from "./length-mm.js";
import { MEASUREMENT_SOURCES } from "./measurement.js";
import { PointMm } from "./point-mm.js";

describe("measurements", () => {
  it("freezes SceneIR 0.1 sources to laser | typed | depth_fit", () => {
    expect([...MEASUREMENT_SOURCES].sort()).toEqual([
      "depth_fit",
      "laser",
      "typed",
    ]);
  });

  it("records a laser length on an opening without dropping source", () => {
    const doc = FloorPlanDocument.create({ id: "doc_1" });
    const storeyId = doc.storeys[0]!.id;
    doc.addWall({
      storeyId,
      id: "wall_s",
      start: PointMm.of(0, 0),
      end: PointMm.of(4000, 0),
      thickness: LengthMm.of(200),
      height: LengthMm.of(2800),
      kind: "exterior",
    });
    doc.addOpening({
      storeyId,
      wallId: "wall_s",
      id: "op_door",
      kind: "door",
      width: LengthMm.of(800),
      height: LengthMm.of(2100),
      offsetAlongWall: LengthMm.of(800),
      sillHeight: LengthMm.of(0),
    });
    doc.setMeasurement({
      id: "m_door_width",
      kind: "length",
      value: LengthMm.of(900),
      source: "laser",
      instrumentId: "laser_sku_x",
      target: { entityType: "opening", entityId: "op_door", field: "width" },
    });
    const measurement = doc.measurements.find((m) => m.id === "m_door_width");
    expect(measurement?.source).toBe("laser");
    expect(measurement?.valueMm).toBe(900);
    expect(
      doc.storeys[0]?.walls
        .find((w) => w.id === "wall_s")
        ?.openings.find((o) => o.id === "op_door")?.width.value,
    ).toBe(900);
  });

  it("rejects an unknown measurement source", () => {
    const doc = FloorPlanDocument.create({ id: "doc_1" });
    expect(() =>
      doc.setMeasurement({
        id: "m_bad",
        kind: "length",
        value: LengthMm.of(1000),
        source: "rf_ble" as "laser",
      }),
    ).toThrow(DomainError);
  });

  it("does not let depth_fit silently overwrite a laser measurement", () => {
    const doc = FloorPlanDocument.create({ id: "doc_1" });
    doc.setMeasurement({
      id: "m_axis",
      kind: "length",
      value: LengthMm.of(4000),
      source: "laser",
      instrumentId: "laser_sku_x",
      between: ["lm_a", "lm_b"],
    });
    expect(() =>
      doc.setMeasurement({
        id: "m_axis",
        kind: "length",
        value: LengthMm.of(4012),
        source: "depth_fit",
      }),
    ).toThrow(/laser/i);
    expect(doc.measurements[0]?.source).toBe("laser");
    expect(doc.measurements[0]?.valueMm).toBe(4000);
  });

  it("allows typed to replace a previous typed value", () => {
    const doc = FloorPlanDocument.create({ id: "doc_1" });
    doc.setMeasurement({
      id: "m_height",
      kind: "length",
      value: LengthMm.of(2700),
      source: "typed",
      target: {
        entityType: "storey",
        entityId: doc.storeys[0]!.id,
        field: "height",
      },
    });
    doc.setMeasurement({
      id: "m_height",
      kind: "length",
      value: LengthMm.of(2800),
      source: "typed",
      target: {
        entityType: "storey",
        entityId: doc.storeys[0]!.id,
        field: "height",
      },
    });
    expect(doc.measurements[0]?.valueMm).toBe(2800);
    expect(doc.storeys[0]?.height.value).toBe(2800);
  });
});

File: core/tests/measurement_test.cpp
#include <gtest/gtest.h>

#include "toporoom/domain/error.hpp"
#include "toporoom/domain/floor_plan_document.hpp"
#include "toporoom/domain/kinds.hpp"
#include "toporoom/domain/length_mm.hpp"
#include "toporoom/domain/point_mm.hpp"

using toporoom::domain::AddOpeningProps;
using toporoom::domain::AddWallProps;
using toporoom::domain::CreateFloorPlanProps;
using toporoom::domain::DomainError;
using toporoom::domain::FloorPlanDocument;
using toporoom::domain::is_measurement_source;
using toporoom::domain::LengthMm;
using toporoom::domain::MeasurementSource;
using toporoom::domain::OpeningKind;
using toporoom::domain::PointMm;
using toporoom::domain::SetMeasurementProps;
using toporoom::domain::WallKind;

TEST(Measurements, FrozenSources) {
  EXPECT_TRUE(is_measurement_source("laser"));
  EXPECT_TRUE(is_measurement_source("typed"));
  EXPECT_TRUE(is_measurement_source("depth_fit"));
  EXPECT_FALSE(is_measurement_source("rf_ble"));
}

TEST(Measurements, RecordsLaserOnOpening) {
  auto doc = FloorPlanDocument::create(CreateFloorPlanProps{"doc_1"});
  const auto storey_id = doc.storeys()[0].id();
  AddWallProps wall;
  wall.storey_id = storey_id;
  wall.id = "wall_s";
  wall.start = PointMm::of(0, 0);
  wall.end = PointMm::of(4000, 0);
  wall.thickness = LengthMm::of(200);
  wall.height = LengthMm::of(2800);
  wall.kind = WallKind::Exterior;
  doc.add_wall(wall);
  AddOpeningProps opening;
  opening.storey_id = storey_id;
  opening.wall_id = "wall_s";
  opening.id = "op_door";
  opening.kind = OpeningKind::Door;
  opening.width = LengthMm::of(800);
  opening.height = LengthMm::of(2100);
  opening.offset_along_wall = LengthMm::of(800);
  doc.add_opening(opening);
  SetMeasurementProps measurement;
  measurement.id = "m_door_width";
  measurement.value = LengthMm::of(900);
  measurement.source = MeasurementSource::Laser;
  measurement.instrument_id = "laser_sku_x";
  measurement.target = {{"opening", "op_door", "width"}};
  doc.set_measurement(measurement);
  ASSERT_FALSE(doc.measurements().empty());
  EXPECT_EQ(doc.measurements()[0].source, MeasurementSource::Laser);
  EXPECT_EQ(doc.measurements()[0].value_mm, 900);
  EXPECT_EQ(doc.storeys()[0].walls()[0].openings()[0].width().value(), 900);
}

TEST(Measurements, DepthFitDoesNotOverwriteLaser) {
  auto doc = FloorPlanDocument::create(CreateFloorPlanProps{"doc_1"});
  SetMeasurementProps laser;
  laser.id = "m_axis";
  laser.value = LengthMm::of(4000);
  laser.source = MeasurementSource::Laser;
  laser.instrument_id = "laser_sku_x";
  laser.between = {"lm_a", "lm_b"};
  doc.set_measurement(laser);
  SetMeasurementProps depth;
  depth.id = "m_axis";
  depth.value = LengthMm::of(4012);
  depth.source = MeasurementSource::DepthFit;
  EXPECT_THROW(doc.set_measurement(depth), DomainError);
  EXPECT_EQ(doc.measurements()[0].source, MeasurementSource::Laser);
  EXPECT_EQ(doc.measurements()[0].value_mm, 4000);
}

TEST(Measurements, TypedCanReplaceTyped) {
  auto doc = FloorPlanDocument::create(CreateFloorPlanProps{"doc_1"});
  SetMeasurementProps first;
  first.id = "m_height";
  first.value = LengthMm::of(2700);
  first.source = MeasurementSource::Typed;
  first.target = {{"storey", doc.storeys()[0].id(), "height"}};
  doc.set_measurement(first);
  SetMeasurementProps second = first;
  second.value = LengthMm::of(2800);
  doc.set_measurement(second);
  EXPECT_EQ(doc.measurements()[0].value_mm, 2800);
  EXPECT_EQ(doc.storeys()[0].height().value(), 2800);
}
